features/64-chat-offline-cache/api.md

```diff
@@ -0,0 +1,373 @@
+- Compatible to old API interface
+
+# Overview
+
+Container contains a local store for caching resources. The local store caches resources by resource id.
+
+### Store Update
+
+When SDK recieves result from server, for each record, the newly fetched record would always replace the old one with the same record id if there is one.
+
+### API
+
+For fetch API of each resource, there would be a new callback that should return result from cache immediately.
+
+The interface for each language may be different.
+
+```
+getResource(params, callback, cachedCallback?)
+```
+
+The cachedCallback should be optional, thus compatible to the old API interace.
+
+After client app receives result from both callback, client app need to resolve a single result for display. Skygear Chat Container would provide some utility functions to help client app resolve two callback result.
+
+### Resolve result
+
+There are two scenarios, and solutions would be provided by SDK. If client app requires custom resolve logic, they need to implement themselves.
+
+#### 1
+
+```
+---
+ ^
+ | older cached result
+ | ---
+ v  ^
+--- |
+    | newly fetched result
+    v
+   ---
+```
+
+Solution:
+
+- Newly fetched result is appended to the older cached result
+- For overlapped records, the newer one is taken
+
+#### 2
+
+```
+---
+ ^
+ | older cached result
+ v
+---
+(a gap bewtween two sets of result)
+---
+ ^
+ | newly fetched result
+ v
+---
+```
+
+Solution:
+
+- Newly fetched result is appended to the older cached result
+
+**How to fill the gap?**
+
+The utility functions we provide should be able to tell developers where the boundaries of the message list are.
+
+```
+let boundaries = skygearChat.findBoundaries(messagelist)
+
+// if no gap in the message list
+// boundaries == [[messageA, messageZ]]
+//
+// app should load messages newer than messageA and older than messageZ
+
+// else
+// boundaries == [[messageA, messageD], [messageG, messageZ]]
+//
+// app should load messages newer than messageA or messageG, and older than messageD or messageZ
+```
+
+This could be quite complicated, so we may provide scroll view delegate or handler implementation to handle this logic for the developer.
+
+# Changes Required
+
+### Server
+
+Add a `previous_message` to `message` table, to determine if two given messages are subsequent messages.
+
+Deleted messages are also returned in fetch API, but the content of the message would be empty. This is to keep the whole message list complete.
+
+### SDK
+
+- Add cached callback to fetch messages and fetch conversations API.
+
+```
+func fetchConversations(cachedCallback?)
+func fetchMessages(conversation, beforeMessage, limit, order, cachedCallback?)
+func fetchMessages(conversation, afterMessage, limit, order, cachedCallback?)
+```
+
+#### JS
+
+```js
+const successCallback = (conversations, cached = false) => {
+  // display conversations
+};
+
+skygearChat
+  .fetchConversations(successCallback)
```

Comment:
I think in js, it is called `getConversations`.

And I think the parameter is a boolean instead of a function, according to the general API signature above?

Reply:
This pattern is very like cached query in js, https://docs.skygear.io/guides/cloud-db/offline-storage/js/.

Because promise can only be resolved once, so the cached result is returned in the function passed as `getConversations` parameter.

Reply:
Though, actually I think `getConversations` is not a good name though, because it is `fetching` resource from server.

Reply:
Oh right Promise can't be resolved twice.... hmmm, is there anyway to do it better? Or some reference to modern JS lib with similar problems? Felt like one callback is via Promise, another one is via function parameter, is quite weird...

Reply:
Like ... return two Promises? (I probably don't know what I'm saying lol)

Reply:
I think `Rx` is common to solve this kind of problem